File: otc/src/com/tsystems/otc/huawei/OtcHttpMethods.java
/* 
 * Copyright (c) 2016 T-Systems GmbH
 * Germany
 * All rights reserved.
 * 
 * Name: ParamFactory.java
 * Author: zsonagy
 * Datum: 08.03.2016
 */

package com.tsystems.otc.huawei;

import java.io.BufferedReader;
import java.io.ByteArrayInputStream;
import java.io.IOException;
import java.io.InputStream;
import java.io.InputStreamReader;
import java.net.MalformedURLException;
import java.net.URL;

import org.apache.http.HttpResponse;
import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;

import com.cloud.sdk.http.HttpMethodName;
import com.tsystems.otc.config.OtcConfig;
import com.tsystems.otc.huawei.AccessService;
import com.tsystems.otc.huawei.AccessServiceImpl;

/***
 * HTTP Methods class to implement GET/POST/PUT/DELETE methods to OTC calls 
 * @author zsonagy
 *
 */
public class OtcHttpMethods {
	
	static final Logger log = LogManager.getLogger(OtcHttpMethods.class.getName());

	public static String put(String requestUrl, String putBody) {
		String ret="";
		
		AccessService accessService = null;
		try {
			accessService = new AccessServiceImpl(OtcConfig.serviceName,
					OtcConfig.region, OtcConfig.sk, OtcConfig.sk);
			URL url = new URL(requestUrl);
			HttpMethodName httpMethod = HttpMethodName.PUT;

			InputStream content = new ByteArrayInputStream(putBody.getBytes());
			HttpResponse response = accessService.access(url, content,
					(long) putBody.getBytes().length, httpMethod);

			ret = convertStreamToString(response.getEntity().getContent());
					//getStatusLine().get();
	
		} catch (Exception e) {
			e.printStackTrace();
		} finally {
			accessService.close();
		}
		return ret;
	}

	public static String patch(String requestUrl, String putBody) {

		String ret=null;
		AccessService accessService = null;
		try {
			accessService = new AccessServiceImpl(OtcConfig.serviceName,
					OtcConfig.region, OtcConfig.ak, OtcConfig.sk);
			URL url = new URL(requestUrl);
			HttpMethodName httpMethod = HttpMethodName.PATCH;
			InputStream content = new ByteArrayInputStream(putBody.getBytes());
			HttpResponse response = accessService.access(url, content,
					(long) putBody.getBytes().length, httpMethod);

			ret = convertStreamToString(response.getEntity()
					.getContent());
		} catch (Exception e) {
			e.printStackTrace();
		} finally {
			accessService.close();
		}

		return ret;
	}

	public static String delete(String requestUrl) {

		String ret = null;
		AccessService accessService = null;

		try {
			accessService = new AccessServiceImpl(OtcConfig.serviceName,
					OtcConfig.region, OtcConfig.ak, OtcConfig.sk);
			URL url = new URL(requestUrl);
			HttpMethodName httpMethod = HttpMethodName.DELETE;

			HttpResponse response = accessService.access(url, httpMethod);
			ret = convertStreamToString(response.getEntity()
					.getContent());
		} catch (Exception e) {
			e.printStackTrace();
		} finally {
			accessService.close();
		}
		return ret;

	}

	public static String get(String requestUrl) {
		String ret = null;

		AccessService accessService = null;

		try {
			accessService = new AccessServiceImpl(OtcConfig.serviceName,
					OtcConfig.region, OtcConfig.ak, OtcConfig.sk);
			URL url = new URL(requestUrl);
			HttpMethodName httpMethod = HttpMethodName.GET;
			HttpResponse response;
			response = accessService.access(url, httpMethod);
			ret = convertStreamToString(response.getEntity().getContent());
		} catch (Exception e) {
			e.printStackTrace();
		} finally {
			accessService.close();
		}

		return ret;
	}

	public static String post(String requestUrl, String postbody) {

		String ret = null;
		AccessService accessService = new AccessServiceImpl(
				OtcConfig.serviceName, OtcConfig.region, OtcConfig.ak,
				OtcConfig.sk);
		URL url = null;
		try {
			url = new URL(requestUrl);
		} catch (MalformedURLException e) {
			e.printStackTrace();
		}
		InputStream content = new ByteArrayInputStream(postbody.getBytes());
		HttpMethodName httpMethod = HttpMethodName.POST;
		HttpResponse response;

		try {
			response = accessService.access(url, content,
					(long) postbody.getBytes().length, httpMethod);
			if( response.getHeaders("X-Subject-Token").length > 0)
			{
				OtcConfig.TOKEN = response.getHeaders("X-Subject-Token")[0].getValue();
			}
						
			ret = convertStreamToString(response.getEntity()
					.getContent());
		} catch (Exception e) {
			e.printStackTrace();
		} finally {
			accessService.close();
		}
		return ret;
	}

	private static String convertStreamToString(InputStream is) {
		BufferedReader reader = new BufferedReader(new InputStreamReader(is));
		StringBuilder sb = new StringBuilder();

		String line = null;
		try {
			while ((line = reader.readLine()) != null) {
				sb.append(line + "\n");
			}
		} catch (IOException e) {
			e.printStackTrace();
		} finally {
			try {
				is.close();
			} catch (IOException e) {
				e.printStackTrace();
			}
		}

		return sb.toString();
	}

}

File: python-otcclient/OtcHttpMethods.py
#!/usr/bin/env python
#
""" generated source for module OtcHttpMethods """
#  
#  * Copyright (c) 2016 T-Systems GmbH
#  * Germany
#  * All rights reserved.
#  * 
#  * Name: ParamFactory.java
#  * Author: zsonagy
#  * Datum: 08.03.2016
#  



from OtcConfig import OtcConfig
import requests
requests.packages.urllib3.disable_warnings()

from wsgiref import headers
from OpenSSL.crypto import verify


class OtcHttpMethods(object):
    """ generated source for class OtcHttpMethods """

    url = 'http://www.someserver.com/cgi-bin/register.cgi'
    user_agent = 'Mozilla/5.0 (Windows NT 6.1; Win64; x64)'
    values = {'name': 'Michael Foord',
              'location': 'Northampton',
              'language': 'Python' }
    headers = {'User-Agent': user_agent}
    
    @classmethod
    def put(cls, requestUrl, putBody):
        """ generated source for method put """
        ret = ""
        accessService = None
        try:
            print "a"
#            accessService = AccessServiceImpl(OtcConfig.serviceName, OtcConfig.region, OtcConfig.sk, OtcConfig.sk)
#            ret = convertStreamToString(response.getEntity().getContent())
            # getStatusLine().get();

        except Exception as e:
            e.printStackTrace()
        finally:
            accessService.close()
        return ret

    @classmethod
    def patch(cls, requestUrl, putBody):
        """ generated source for method patch """
        ret = None
        accessService = None
        try:
            print "a"
#            accessService = AccessServiceImpl(OtcConfig.serviceName, OtcConfig.region, OtcConfig.ak, OtcConfig.sk)
#            ret = convertStreamToString(response.getEntity().getContent())
        except Exception as e:
            e.printStackTrace()
        finally:
            accessService.close()
        return ret

    @classmethod
    def delete(cls, requestUrl):
        """ generated source for method delete """
        ret = None
        accessService = None 
        try:
            print "a"
#            accessService = AccessServiceImpl(OtcConfig.serviceName, OtcConfig.region, OtcConfig.ak, OtcConfig.sk)
#            ret = convertStreamToString(response.getEntity().getContent())
        except Exception as e:
            e.printStackTrace()
        finally:
            accessService.close()
        return ret

    @classmethod
    def get(cls, requestUrl):
        """ generated source for method post """
        ret = None
#        accessService = AccessServiceImpl(OtcConfig.serviceName, OtcConfig.region, OtcConfig.ak, OtcConfig.sk)
        try:
            response = cls.httpcall(requestUrl)  
                     
            token = response.headers.get('X-Subject-Token')   
            if  token != None and len( response.headers.get('X-Subject-Token')) > 0:
                OtcConfig.TOKEN = response.headers.get("X-Subject-Token")
            
            ret = response.text
        except Exception as e:
            print e
        finally:
            pass
            #print "TODO close"
        return ret


    @classmethod
    def post(cls, requestUrl, postbody):
        """ generated source for method post """
        ret = None
#        accessService = AccessServiceImpl(OtcConfig.serviceName, OtcConfig.region, OtcConfig.ak, OtcConfig.sk)
        try:
            response = cls.httpcall(requestUrl, datastr=str(postbody))  
                        
            token = response.headers.get('X-Subject-Token')   
            if  token != None and len( response.headers.get('X-Subject-Token')) > 0:
                OtcConfig.TOKEN = response.headers.get("X-Subject-Token")            
            ret = response.text
        except Exception as e:
            print e
        finally:
            pass
            #print "TODO close"
        return ret

    
    @classmethod
    def httpcall( cls,url, datastr=None):    
        
        s = requests.session()
        
#After that, continue with your requests as you would:

#logged in! cookies saved for future requests.

        headers = {'Content-Type': 'application/json',  'Accept': 'application/json' }
        
        if len(OtcConfig.TOKEN) > 0:
            headers['X-Auth-Token'] = OtcConfig.TOKEN
        if datastr:
            data = datastr #urllib.urlencode(datastr)
            #print data
            response=s.post(url, data, headers=headers, verify=False)
        else:
            response=s.get(url, headers=headers, verify=False)
        return response
#proxy        
        #proxy_support = urllib2.ProxyHandler({})
        #opener = urllib2.build_opener(proxy_support)
        #urllib2.install_opener(opener)

    @classmethod
    def convertStreamToString(cls, is_):
        """ generated source for method convertStreamToString """
